### ISSUE TITLE
识别长音频问题

### ISSUE THREAD

**Issue by lwppwl** [yeyupiaoling/Whisper-Finetune#5]
你好，我按照你的文档教程，使用large-v2模型在aishell数据集上训练得到了一个模型，然后使用infer.py识别一个2min45s的长音频时，只输出了第一句话。使用infer_ct.py时也只输出了第一句话和中间的2句话。请问这是什么问题呢？我用是5月11号的代码，最新代码的 max_audio_len参数能否解决这个问题呢？

**Comment by yeyupiaoling** [comment]
你这个音频软件适合发给我吗？我这边也测试一下。

**Comment by lwppwl** [comment]
[演示音频.zip](https://github.com/yeyupiaoling/Whisper-Finetune/files/11477010/default.zip)
这个是2min45s的长音频

**Comment by starry-zfr** [comment]
我也遇到了这个问题，识别长音频时只识别出前几个字

**Comment by yeyupiaoling** [comment]
@lwppwl 你的文件识别结果，正常啊！

```
{
    "language": "zh",
    "duration": 165.9713125,
    "results": [
        {
            "start": 0.0,
            "end": 4.0,
            "text": "欢迎使用闪电音频格式转换器"
        },
        {
            "start": 4.0,
            "end": 7.0,
            "text": "您身边好用的音频处理助手"
        },
        {
            "start": 7.0,
            "end": 9.0,
            "text": "实验名称"
        },
        {
            "start": 9.0,
            "end": 13.0,
            "text": "流逝细胞仪检测细胞凋亡的流程"
        },
        {
            "start": 13.0,
            "end": 14.0,
            "text": "操作步骤"
        },
        {
            "start": 14.0,
            "end": 15.0,
            "text": "第一步"
        },
        {
            "start": 15.0,
            "end": 17.0,
            "text": "取Falcon式管"
        },
        {
            "start": 17.0,
            "end": 22.0,
            "text": "按标本顺序编好音性对照管和标本管号"
        },
        {
            "start": 22.0,
            "end": 23.0,
            "text": "第二步"
        },
        {
            "start": 23.0,
            "end": 28.0,
            "text": "使用冷的PBS缓冲液洗细胞两次"
        },
        {
            "start": 29.0,
            "end": 38.0,
            "text": "再用1X binding buffer缓冲液配成1X10的六次方细胞每ml的旋液"
        },
        {
            "start": 38.0,
            "end": 39.0,
            "text": "第三步"
        },
        {
            "start": 39.0,
            "end": 46.0,
            "text": "取Falcon式管中加入100mL的细胞旋液"
        },
        {
            "start": 46.0,
            "end": 47.0,
            "text": "第四步"
        },
        {
            "start": 47.0,
            "end": 54.0,
            "text": "按以下体积加入Anaxin V与核酸染料"
        },
        {
            "start": 55.0,
            "end": 57.0,
            "text": "实验过程编号"
        },
        {
            "start": 57.0,
            "end": 58.0,
            "text": "一"
        },
        {
            "start": 58.0,
            "end": 59.0,
            "text": "音性对照"
        },
        {
            "start": 59.0,
            "end": 63.0,
            "text": "不加入荧光标记Anaxin V"
        },
        {
            "start": 63.0,
            "end": 65.0,
            "text": "不加入核酸染料"
        },
        {
            "start": 65.0,
            "end": 66.0,
            "text": "第二步"
        },
        {
            "start": 66.0,
            "end": 67.0,
            "text": "单阳一"
        },
        {
            "start": 67.0,
            "end": 70.0,
            "text": "加入AVFITC"
        },
        {
            "start": 70.0,
            "end": 73.0,
            "text": "不加入核酸染料"
        },
        {
            "start": 73.0,
            "end": 74.0,
            "text": "第三步"
        },
        {
            "start": 74.0,
            "end": 75.0,
            "text": "单阳二"
        },
        {
            "start": 75.0,
            "end": 79.0,
            "text": "不加入荧光标记Anaxin V"
        },
        {
            "start": 79.0,
            "end": 81.0,
            "text": "加入核酸染料PI"
        },
        {
            "start": 81.0,
            "end": 82.0,
            "text": "第四步"
        },
        {
            "start": 82.0,
            "end": 83.0,
            "text": "样本"
        },
        {
            "start": 83.0,
            "end": 87.0,
            "text": "加入AVFITC"
        },
        {
            "start": 87.0,
            "end": 90.0,
            "text": "不加入核酸染料PI"
        },
        {
            "start": 90.0,
            "end": 91.0,
            "text": "第五步"
        },
        {
            "start": 91.0,
            "end": 95.0,
            "text": "轻轻混匀室温20-25度"
        },
        {
            "start": 95.0,
            "end": 100.0,
            "text": "闭光处放置15分钟左右"
        },
        {
            "start": 100.0,
            "end": 101.0,
            "text": "第六步"
        },
        {
            "start": 101.0,
            "end": 107.0,
            "text": "使用Anaxin V Biotin试剂进行检测时"
        },
        {
            "start": 107.0,
            "end": 113.0,
            "text": "一乘以斑丁巴富尔缓冲液洗细胞一次去上清"
        },
        {
            "start": 113.0,
            "end": 127.0,
            "text": "一乘斑丁巴富尔缓冲液100µL溶解SAV-FITC试剂0.5微克加入到细胞管中"
        },
        {
            "start": 128.0,
            "end": 129.0,
            "text": "轻轻混匀"
        },
        {
            "start": 129.0,
            "end": 133.0,
            "text": "加入5µL PI"
        },
        {
            "start": 133.0,
            "end": 136.0,
            "text": "室温20-25摄氏度"
        },
        {
            "start": 136.0,
            "end": 140.0,
            "text": "闭光处放置15分钟"
        },
        {
            "start": 140.0,
            "end": 141.0,
            "text": "七"
        },
        {
            "start": 141.0,
            "end": 146.0,
            "text": "各室管中分别加入一成的斑丁巴富尔缓冲液"
        },
        {
            "start": 146.0,
            "end": 148.0,
            "text": "400µL"
        },
        {
            "start": 148.0,
            "end": 155.0,
            "text": "把一小时内上流逝细胞仪测定结果"
        },
        {
            "start": 155.0,
            "end": 158.0,
            "text": "实验结束"
        }
    ],
    "text": "欢迎使用闪电音频格式转换器您身边好用的音频处理助手实验名称流逝细胞仪检测细胞凋亡的流程操作步骤第一步取Falcon式管按标本顺序编好音性对照管和标本管号第二步使用冷的PBS缓冲液洗细胞两次再用1X binding buffer缓冲液配成1X10的六次方细胞每ml的旋液第三步取Falcon式管中加入100mL的细胞旋液第四步按以下体积加入Anaxin V与核酸染料实验过程编号一音性对照不加入荧光标记Anaxin V不加入核酸染料第二步单阳一加入AVFITC不加入核酸染料第三步单阳二不加入荧光标记Anaxin V加入核酸染料PI第四步样本加入AVFITC不加入核酸染料PI第五步轻轻混匀室温20-25度闭光处放置15分钟左右第六步使用Anaxin V Biotin试剂进行检测时一乘以斑丁巴富尔缓冲液洗细胞一次去上清一乘斑丁巴富尔缓冲液100µL溶解SAV-FITC试剂0.5微克加入到细胞管中轻轻混匀加入5µL PI室温20-25摄氏度闭光处放置15分钟七各室管中分别加入一成的斑丁巴富尔缓冲液400µL把一小时内上流逝细胞仪测定结果实验结束"
}
```

**Comment by lwppwl** [comment]
@yeyupiaoling  请问使用的是large-v2微调后的模型吗？原模型我测试也是没问题的，微调后的模型后的模型才有的这个问题。
可能是我没用最新的代码的问题吗？下面是我当时微调时所用的参数：
base_model: openai/whisper-large-v2
eval_steps: 10000
generation_max_length: 128
gradient_accumulation_steps: 2
language: Chinese
learning_rate: 0.001
logging_steps: 100
num_train_epochs: 3
num_workers: 8
output_path: models/whisper-large-v2-lora
per_device_eval_batch_size: 2
per_device_train_batch_size: 2
resume_from_checkpoint: None
save_steps: 10000
task: transcribe
test_data: dataset/test.json
train_data: dataset/train.json
use_8bit: False
warmup_steps: 50

**Comment by yeyupiaoling** [comment]
我知道你的问题了，你用的是infer.py，这个是没有长音频处理功能的，你要用infer_ct2.py

**Comment by lwppwl** [comment]
@yeyupiaoling  使用infer_ct2.py也没有全部识别。可能我需要用最新的代码重新训练一下。

**Comment by lwppwl** [comment]
@yeyupiaoling  大佬，我训练后large-v2模型的识别结果全为!!!!!!!!，这是什么原因呢？单卡V100-32G，因为显存不够，我把per_device_train_batch_size和per_device_eval_batch_size调为1 ， 具体训练参数：
![20230522-134917](https://github.com/yeyupiaoling/Whisper-Finetune/assets/45410895/3026ba5a-5730-45b8-8994-3eeb02335dab)

**Comment by lwppwl** [comment]
@yeyupiaoling  tensorboard上看，训练的并不好，有什么参数修改意见吗？
![20230522-135544](https://github.com/yeyupiaoling/Whisper-Finetune/assets/45410895/4916a65f-8d2a-42f7-a7c6-a3757d1035a1)

**Comment by yeyupiaoling** [comment]
@lwppwl 这个我也看不出来，你的训练音频长度是多少的？是用的最新代码吗？per_device_train_batch_size最好大于等于2，可以使用8bit训练，显存就够了

**Comment by lwppwl** [comment]
@yeyupiaoling 使用tiny模型训练，设置batch_size=32，输出正常了。但是我有个疑问，每行输出的时间段都是30秒，finetune时需要改什么参数吗？下面两张图分别是我训练的tiny和你发布的的tiny模型的输出结果。
![20230523-114107](https://github.com/yeyupiaoling/Whisper-Finetune/assets/45410895/5b7c57e4-56b8-48f3-93c4-eba0ec7d0359)
![20230523-114112](https://github.com/yeyupiaoling/Whisper-Finetune/assets/45410895/1a8c69c5-3590-4cef-8f6e-2a4a61236317)

**Comment by lwppwl** [comment]
@yeyupiaoling  使用tiny模型训练，设置batch_size=32，输出正常了。但是我有个疑问，每行输出的时间段都是30秒，是不是训练的不够充分？
![20230523-114107](https://github.com/yeyupiaoling/Whisper-Finetune/assets/45410895/da3a7718-6e0b-4d55-933d-7044b662a468)

**Comment by yeyupiaoling** [comment]
用的是哪个脚本预测。

**Comment by lwppwl** [comment]
@yeyupiaoling infer_ct2.py ，使用的是tiny模型

**Comment by cyc00518** [comment]
@yeyupiaoling 
微調後的tiny模型(batch_size=8)
也有相同的問題, 超過30秒的部分會自動忽略, 且每行輸出都是以30秒

small模型則沒有此問題。

**Comment by yeyupiaoling** [comment]
>使用tiny模型训练，设置batch_size=32，输出正常了。但是我有个疑问，每行输出的时间段都是30秒，是不是训练的不够充分？

@lwppwl 微调应该不影响分割的，我也没有试过微调后推理长音频。我有空试试

**Comment by yeyupiaoling** [comment]
>微調後的tiny模型(batch_size=8)
>也有相同的問題, 超過30秒的部分會自動忽略, 且每行輸出都是以30秒
>
>small模型則沒有此問題。

@cyc00518 有试过用原始模型吗？就是没有微调的也会出现这种情况吗？

**Comment by cyc00518** [comment]
@yeyupiaoling 原始模型都沒有問題, 只有微調過後的會出現這種情況。
我微調的訓練範本是Hugging Face 的CommonVoice 12的訓練資料, 拿前6000筆資料來訓練,
訓練參數如下:
`training_args = Seq2SeqTrainingArguments(
    output_dir="int8-whisper-tiny-chinese",
    per_device_train_batch_size=8,
    gradient_accumulation_steps=1,
    learning_rate=1e-3,
    warmup_steps=50,
    num_train_epochs=20,
    evaluation_strategy="epoch", 
    fp16=True,
    per_device_eval_batch_size=8,
    generation_max_length=128,
    logging_steps=100,
    remove_unused_columns=False,
    label_names=["labels"],
)`

**Comment by yeyupiaoling** [comment]
参数看着没有问题。

**Comment by yeyupiaoling** [comment]
@cyc00518 现在支持微调带时间戳数据的，这样微调后的模型，应该不会出现你说的只分割30秒的情况

**Comment by wwfcnu** [comment]
whisper原本的训练代码开源了吗
